### DISCUSSION TITLE
Asking questions to a Specific Documents

### DISCUSSION THREAD

**Discussion by sandeepanprime** [langroid/langroid#430] [Q&A]
Hi, 

We have a use case where we have got large documents running into 500+ pages. Let's say we have 10 of them for simplicity (Doc1, Doc2,..., Doc10). Now when a question comes, the question comes along with the document ID (DocN). How do I use Langroid to run the RAG only on that document? I want to be able to ingest all the documents but run question (query) only on a specific document.

**Comment by pchalasani** [comment]
The best way is to include the doc id in the metadata during ingestion. You can use the `ingest_doc_paths` (or similar)  methods in `DocChatAgent` which accept a `metadata` arg. When the docs are chunked, the parent doc metadata is duplicated across all chunks within the doc (so, e.g. all of Document 24's chunks will have the same `metadata.doc_id = 24`).
Once you've ingested this way, you can define a tool , say `FilteredQuery`, with fields `filter` and `query`, instruct the LLM to use this tool, write a handler method to use the `filter` to set the `config.filter` to restrict the doc-set so that RAG will only use this.

Some pointers for you to take a look -- 

https://github.com/langroid/langroid/blob/main/examples/docqa/filter-multi-doc-auto.py

https://github.com/langroid/langroid/blob/main/langroid/agent/special/doc_chat_agent.py#L939

A pre-defined Filtered rag is available if you use lancedb, see this example https://github.com/langroid/langroid/blob/main/examples/docqa/lance-rag-movies.py
It uses the `LanceRAGTaskCreator`:

https://github.com/langroid/langroid/blob/main/langroid/agent/special/lance_rag/lance_rag_task.py

**Comment by pchalasani** [comment]
@sandeepanprime I made another example script closer to what you're looking for:
https://github.com/langroid/langroid/blob/main/examples/docqa/filter-multi-doc-query-plan.py

**Comment by sandeepanprime** [comment]
Thank you Prasad. Will try and let you know. Exactly what I was looking for

**Comment by carl-mate** [comment]
@pchalasani Hi Prasad, I appreciate your work with Langroid.

I'm new to this package (and RAG, in general) and I found this thread very helpful for document filtering, especially the example you gave: https://github.com/langroid/langroid/blob/main/examples/docqa/filter-multi-doc-query-plan.py

I would like to ask your help or maybe some hints on the document ingestion part:

```
            agent.ingest_doc_paths(
                [paths[musician]],  # all chunks of this doc will have same metadata
                metadata[musician],
            )
```

I have tried the code and it works perfectly. However, the page metadata for each chunk is not being included in the extracts. I thought page data would automatically be embedded using the ParsingConfig, but this doesn't seem to be the case.

Can you help me figure out how to include the page for each chunk in the metadata? Thanks!

**Reply by pchalasani** [reply]
> However, the page metadata for each chunk is not being included in the extracts. I thought page data would automatically be embedded using the ParsingConfig, but this doesn't seem to be the case.

Thanks @carl-mate,  can you elaborate ? The `Document` class has a field `metadata` of class `DocMetaData` which you have to populate as needed. Some fields like `source` are filled in with default values, but any other metadata, you have to insert using your own code. 

E.g., we insert some metadata here:
https://github.com/langroid/langroid/blob/main/examples/quick-start/chat-agent-docs.py

> how to include the page for each chunk in the metadata?

I didn't understand what you meant by "how to include the Page" - did you mean "page number"? Our Pdf parsers do automatically create page numbers, which are included as part of the `metadata.source` field during chunking.

**Reply by carl-mate** [reply]
@pchalasani 

> did you mean "page number"? 


Yes. For further context, I experimented with the [FilterDocAgent](https://github.com/langroid/langroid/blob/22775e68d9d4b4dc8678674e445f15a9c56ebb7d/examples/docqa/filter-multi-doc-query-plan.py#L60-L94) function by printing the out the `Extracts` just to check what is being sent to the LLM.

What I found is that adding "custom" metadata during ingestion:

```
            agent.ingest_doc_paths(
                [paths[document_type]],  
                metadata[document_type], <-- adding custom metadata
            )
```

Seems to override the default `metadata.source` as seen in below example:

```
Extracts: 

 Document(content='....', metadata=DocMetaData(source='context', is_chunk=True, id='a7e08c1f-1a0c-4567-9db9-878f79a79de5', window_ids=['d4c99a29-5aa5-404e-89fe-edb7e5df627c'], document_type='specification'))]
```
 
 Notice how `source='context'`
 
 However, when I don't add any "custom" metadata:
 
 ```
            agent.ingest_doc_paths(
                [paths[document_type]], 
            )
```

And temporarily comment out the filters in the `FilterDocAgent` just to observe `Extracts`

```
class FilterDocAgent(DocChatAgent):
    def llm_response(self, message: None | str | ChatDocument = None) -> Optional[ChatDocument]:
        return ChatAgent.llm_response(self, message)

    def query_plan(self, message: QueryPlanTool) -> str:
        # document_filter_dict = dict(
        #     should=[dict(key="metadata.document_type", match=dict(value=message.document_type))],
        # )
        # 
        # document_filter = json.dumps(document_filter_dict)
        # with temp_update(self.config, {"filter": document_filter}):
        #     self.setup_documents(filter=document_filter)
        #     extracts = self.get_relevant_chunks(message.query)
        self.setup_documents()
        extracts = self.get_relevant_chunks(message.query)

       print('Extracts: ', extracts)

        prompt = f"""
        Answer the QUESTION below based on the following EXTRACTS:
        
        EXTRACTS:
        {extracts}
        
        QUESTION: {message.query}
        """
        response = ChatAgent.llm_response(self, prompt)
        return response.content
```

The `Extracts` will look like this:

`Document(content='...', metadata=DocMetaData(source='/projectFiles/Test Project/specs/specifications.pdf pages 1242-1242', is_chunk=True, id='5188f3b0-1b5c-4512-93cf-dc481edf63b7', window_ids=['c266d02e-27d8-48f8-ab5e-8e8fa45dc772', '51dae8e0-3892-4226-88b5-76f9958bf940', 'c5751958-0e9c-4dcb-bf5a-57193eaa2a66', 'c341906c-ea25-4ea4-a834-46f51bf6d8eb', '5e1843bb-382f-4e74-a0c5-08377a0501ab']))]`

Notice how `source='/projectFiles/Test Project/specs/specifications.pdf pages 1242-1242'`
The `source` metadata is important in my use case because it will allow users to double check the LLM's references from their responses.

In summary, adding "custom" metadata seems to override the `source` in `Extracts`. It would be nice to somehow "merge" the "custom" metadata with the default metadata that was generated by the pdf parsers.

 I would appreciate any guidance and hints. Thank you for taking the time to answer my question!

**Reply by pchalasani** [reply]
Thanks for putting together all this detail! So basically, when providing explicit `metadata` in the `ingest_doc*` methods, you want the `source` part of the provided metadata to be preserved and concatenated with the provided `metadata.source` ?

**Reply by carl-mate** [reply]
@pchalasani Yes, precisely!

**Reply by pchalasani** [reply]
> @pchalasani Yes, precisely!

It's a good idea, we'll release that today, as part of this PR #692

**Reply by carl-mate** [reply]
That's wonderful news. Thank you!

**Reply by pchalasani** [reply]
please get v0.35.0
